x-pack/plugins/reporting/server/export_types/csv_searchsource/generate_csv/generate_csv.ts

```diff
@@ -0,0 +1,384 @@
+      return {
+        size: 0,
+        content: '',
+        maxSizeReached: true,
+        warnings: [],
+      };
+    }
+  }
+
+  /*
+   * Format a Datatable into rows of CSV content
+   */
+  private generateRows(
+    fields: SearchFieldValue[] | undefined,
+    table: Datatable,
+    builder: MaxSizeStringBuilder,
+    formatters: Record<string, FieldFormat>,
+    settings: CsvExportSettings
+  ) {
+    // write the rows
+    this.logger.debug(`Building ${table.rows.length} CSV data rows...`);
+    const { checkForFormulas, escapeValue, separator } = settings;
+
+    for (const dataTableRow of table.rows) {
+      if (this.cancellationToken.isCancelled()) {
+        break;
+      }
+
+      const columnMap = this.getColumnMap(fields, table);
+      const row =
+        columnMap
+          .map((columnIndex, position) => {
+            const tableColumn = table.columns[columnIndex];
+            let cell: string[] | string = '-';
+
+            if (tableColumn != null) {
+              cell = formatters[tableColumn.id].convert(dataTableRow[tableColumn.id]);
+
+              try {
+                // expected values are a string of JSON where the value(s) is in an array
+                cell = JSON.parse(cell);
+              } catch (e) {
+                // ignore
+              }
+
+              // We have to strip singular array values out of their array wrapper,
+              // So that the value appears the visually the same as seen in Discover
+              if (Array.isArray(cell)) {
+                cell = cell.join(', '); // mimic Discover behavior
+              }
+            } else {
+              this.logger.warn(`Unrecognized field: ${(fields && fields[position]) || 'unknown'}`);
+            }
+
+            return cell;
+          })
+          .map((value) => {
+            if (checkForFormulas && cellHasFormulas(value)) {
+              this.csvContainsFormulas = true; // set warning if cell value has a formula
+            }
+            // Escape the values in Data
+            return escapeValue(value);
+          })
+          .join(separator) + '\n';
+
+      if (!builder.tryAppend(row)) {
+        this.logger.warn('max Size Reached');
+        this.maxSizeReached = true;
+        if (this.cancellationToken) {
+          this.cancellationToken.cancel();
+        }
+        break;
+      }
+
+      this.csvRowCount++;
+    }
+  }
+
+  public async generateData(): Promise<TaskRunResult> {
+    const [settings, searchSource] = await Promise.all([
+      getExportSettings(
+        this.clients.uiSettings,
+        this.config,
+        this.job.browserTimezone,
+        this.logger
+      ),
+      this.dependencies.searchSourceStart.create(this.job.searchSource),
+    ]);
+
+    const index = searchSource.getField('index');
+
+    if (!index) {
+      throw new Error(`The search must have a revference to an index pattern!`);
+    }
+
+    const fields = searchSource.getField('fields');
```

Comment:
don't use this, it might be missing stuff (fieldsFromSource for example), rather use the `datatable.columns.map(column => column.name)`

Reply:
On the browser side, `getSharingData` makes sure that the search source `fields` are prepared correctly: https://github.com/tsullivan/kibana/blob/reporting/searchsource-exporttype/src/plugins/discover/public/application/helpers/get_sharing_data.ts#L36:L36

I think it's important to abstract `fieldsFromSource` out on the integration side, since that option is deprecated.

I ran a quick test using `fieldsFromSource` in Advanced Settings, and things seem to work fine without changing the code. Let me know if I am missing something.

> `datatable.columns.map(column => column.name)`

It's more preferable to use `searchSource.getField('fields')` here since it is populated with the columns that are selected in Discover: https://github.com/tsullivan/kibana/blob/reporting/searchsource-exporttype/src/plugins/discover/public/application/helpers/get_sharing_data.ts#L62